scarb/src/core/registry/client/cache.rs

```diff
@@ -139,14 +139,16 @@ impl<'c> RegistryClientCache<'c> {
     pub async fn download_with_cache_and_verify_checksum(
         &self,
         package: PackageId,
-    ) -> Result<FileLockGuard> {
+    ) -> Result<(FileLockGuard, Checksum)> {
         // Skip downloading if the package already has been.
         if self.is_package_downloaded(package).await {
             trace!("found cached archive which is not empty, skipping download");
             let tarball_name = package.tarball_name();
-            return self
+            let file = self
                 .dl_fs
-                .open_rw(&tarball_name, &tarball_name, self.config);
+                .open_rw(&tarball_name, &tarball_name, self.config)?;
+            let checksum = self.get_record_maybe_uncached(package).await?.checksum;
```

Comment:
This may be personal impression, but I would find it cleaner if this has been called only once for both cases (instead of hiding alternative call in `verify_checksum` function). The `verify_checksum` wouldn't even need to take `self` then, this could be a separate struct all together.

Reply:
Note that both logic branches diverge before calling `get_record_maybe_uncached` + there are logic branches which do not need this call. This means, that extracting this value would be suboptimal, as for some branches, the computation result would be discarded.

I did a bit of rewiring in #834 though.